File: packages/coding-agent/src/core/slash-commands.ts
import { t } from "@earendil-works/pi-tui";
import { APP_NAME } from "../config.ts";
import type { SourceInfo } from "./source-info.ts";

export type SlashCommandSource = "extension" | "prompt" | "skill";

export interface SlashCommandInfo {
	name: string;
	description?: string;
	source: SlashCommandSource;
	sourceInfo: SourceInfo;
}

export interface BuiltinSlashCommand {
	name: string;
	description: string;
	argumentHint?: string;
}

/**
 * Get built-in slash commands with localized descriptions.
 * Call this function after i18n is initialized.
 */
export function getBuiltinSlashCommands(): ReadonlyArray<BuiltinSlashCommand> {
	return [
		{ name: "settings", description: t("codingAgent.slash.settings") },
		{ name: "model", description: t("codingAgent.slash.model"), argumentHint: "<provider/model>" },
		{ name: "scoped-models", description: t("codingAgent.slash.scopedModels") },
		{ name: "export", description: t("codingAgent.slash.export") },
		{ name: "import", description: t("codingAgent.slash.import") },
		{ name: "share", description: t("codingAgent.slash.share") },
		{ name: "copy", description: t("codingAgent.slash.copy") },
		{ name: "name", description: t("codingAgent.slash.name") },
		{ name: "session", description: t("codingAgent.slash.session") },
		{ name: "changelog", description: t("codingAgent.slash.changelog") },
		{ name: "hotkeys", description: t("codingAgent.slash.hotkeys") },
		{ name: "fork", description: t("codingAgent.slash.fork") },
		{ name: "clone", description: t("codingAgent.slash.clone") },
		{ name: "tree", description: t("codingAgent.slash.tree") },
		{ name: "trust", description: t("codingAgent.slash.trust") },
		{ name: "login", description: t("codingAgent.slash.login"), argumentHint: "<provider>" },
		{ name: "logout", description: t("codingAgent.slash.logout") },
		{ name: "new", description: t("codingAgent.slash.new") },
		{ name: "compact", description: t("codingAgent.slash.compact") },
		{ name: "resume", description: t("codingAgent.slash.resume") },
		{ name: "reload", description: t("codingAgent.slash.reload") },
		{ name: "quit", description: t("codingAgent.slash.quit", { appName: APP_NAME }) },
	];
}

// Re-export for backward compatibility (non-localized, uses English keys)
export const BUILTIN_SLASH_COMMANDS: ReadonlyArray<BuiltinSlashCommand> = [
	{ name: "settings", description: t("codingAgent.slashFallback.settings") },
	{ name: "model", description: t("codingAgent.slashFallback.model"), argumentHint: "<provider/model>" },
	{ name: "scoped-models", description: t("codingAgent.slashFallback.scopedModels") },
	{ name: "export", description: t("codingAgent.slashFallback.export") },
	{ name: "import", description: t("codingAgent.slashFallback.import") },
	{ name: "share", description: t("codingAgent.slashFallback.share") },
	{ name: "copy", description: t("codingAgent.slashFallback.copy") },
	{ name: "name", description: t("codingAgent.slashFallback.name") },
	{ name: "session", description: t("codingAgent.slashFallback.session") },
	{ name: "changelog", description: t("codingAgent.slashFallback.changelog") },
	{ name: "hotkeys", description: t("codingAgent.slashFallback.hotkeys") },
	{ name: "fork", description: t("codingAgent.slashFallback.fork") },
	{ name: "clone", description: t("codingAgent.slashFallback.clone") },
	{ name: "tree", description: t("codingAgent.slashFallback.tree") },
	{ name: "trust", description: t("codingAgent.slashFallback.trust") },
	{ name: "login", description: t("codingAgent.slashFallback.login"), argumentHint: "<provider>" },
	{ name: "logout", description: t("codingAgent.slashFallback.logout") },
	{ name: "new", description: t("codingAgent.slashFallback.newSession") },
	{ name: "compact", description: t("codingAgent.slashFallback.compact") },
	{ name: "resume", description: t("codingAgent.slashFallback.resume") },
	{ name: "reload", description: t("codingAgent.slashFallback.reload") },
	{ name: "quit", description: t("codingAgent.slashFallback.quit", { appName: APP_NAME }) },
];

File: python/src/pi_mono/coding_agent/core/slash_commands.py
"""Built-in slash commands for the coding agent."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from pi_mono.config import APP_NAME
from pi_mono.coding_agent.core.source_info import SourceInfo

SlashCommandSource = Literal["extension", "prompt", "skill"]


@dataclass(frozen=True)
class SlashCommandInfo:
    name: str
    source: SlashCommandSource
    source_info: SourceInfo
    description: str | None = None


@dataclass(frozen=True)
class BuiltinSlashCommand:
    name: str
    description: str


BUILTIN_SLASH_COMMANDS: tuple[BuiltinSlashCommand, ...] = (
    BuiltinSlashCommand("settings", "Open settings menu"),
    BuiltinSlashCommand("model", "Select model (opens selector UI)"),
    BuiltinSlashCommand("scoped-models", "Enable/disable models for Ctrl+P cycling"),
    BuiltinSlashCommand("export", "Export session (HTML default, or specify path: .html/.jsonl)"),
    BuiltinSlashCommand("import", "Import and resume a session from a JSONL file"),
    BuiltinSlashCommand("share", "Share session as a secret GitHub gist"),
    BuiltinSlashCommand("copy", "Copy last agent message to clipboard"),
    BuiltinSlashCommand("name", "Set session display name"),
    BuiltinSlashCommand("session", "Show session info and stats"),
    BuiltinSlashCommand("changelog", "Show changelog entries"),
    BuiltinSlashCommand("hotkeys", "Show all keyboard shortcuts"),
    BuiltinSlashCommand("fork", "Create a new fork from a previous user message"),
    BuiltinSlashCommand("clone", "Duplicate the current session at the current position"),
    BuiltinSlashCommand("tree", "Navigate session tree (switch branches)"),
    BuiltinSlashCommand("trust", "Change project trust decision"),
    BuiltinSlashCommand("login", "Configure provider authentication"),
    BuiltinSlashCommand("logout", "Remove provider authentication"),
    BuiltinSlashCommand("new", "Start a new session"),
    BuiltinSlashCommand("compact", "Manually compact the session context"),
    BuiltinSlashCommand("resume", "Resume a different session"),
    BuiltinSlashCommand("reload", "Reload keybindings, extensions, skills, prompts, and themes"),
    BuiltinSlashCommand("quit", f"Quit {APP_NAME}"),
)
